OpenRA.Mods.Cnc/Traits/MadTank.cs

```diff
@@ -215,15 +147,117 @@ void IResolveOrder.ResolveOrder(Actor self, Order order)
 					self.CancelActivity();
 
 				self.SetTargetLine(order.Target, Color.Red);
-				self.QueueActivity(new MoveAdjacentTo(self, order.Target, targetLineColor: Color.Red));
-				self.QueueActivity(new CallFunc(StartDetonationSequence));
+				self.QueueActivity(new DetonationSequence(self, this, order.Target));
 			}
 			else if (order.OrderString == "Detonate")
 			{
 				if (!order.Queued)
 					self.CancelActivity();
 
-				self.QueueActivity(new CallFunc(StartDetonationSequence));
+				self.QueueActivity(new DetonationSequence(self, this));
+			}
+		}
+
+		class DetonationSequence : Activity
+		{
+			readonly Actor self;
+			readonly MadTank mad;
+			readonly WithFacingSpriteBody wfsb;
+			readonly ScreenShaker screenShaker;
+			readonly bool assignTargetOnFirstRun;
+
+			int ticks;
+			bool initiated;
+			Target target;
+
+			public DetonationSequence(Actor self, MadTank mad)
+				: this(self, mad, Target.Invalid)
+			{
+				assignTargetOnFirstRun = true;
+			}
+
+			public DetonationSequence(Actor self, MadTank mad, Target target)
+			{
+				this.self = self;
+				this.mad = mad;
+				this.target = target;
+				wfsb = self.Trait<WithFacingSpriteBody>();
+				screenShaker = self.World.WorldActor.Trait<ScreenShaker>();
+			}
+
+			protected override void OnFirstRun(Actor self)
+			{
+				if (assignTargetOnFirstRun)
+					target = Target.FromCell(self.World, self.Location);
+				else
+					QueueChild(new MoveAdjacentTo(self, target, targetLineColor: Color.Red));
```

Comment:
Unfortunately this still isn't quite right :(

Repro case: Target an enemy building, then place a wall or another blocking structure to interrupt the path of the MAD tank. It will deploy at the location it is blocked instead of repathing to the target.

Moving this into `Tick` like the other activities do should fix this.

Reply:
Hopefully got it right this time...